test/async-action.ts

```diff
@@ -214,6 +214,192 @@ test("it should support logging", t => {
     }, 10)
 })
 
+test("it should support asyncActionWithCancel(cb, type) as decorator for have-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        *f(initial: number) {
+            this.a = initial // this runs in action
+            try {
+                this.a = yield delay(100, 5, true) // and this as well!
+                yield delay(100, 0)
+                this.a = 4
+            } catch (e) {
+                this.a = e
+            }
+            return this.a
+        }
+    }
+
+    const x = new X()
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.f(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support asyncActionWithCancel(cb, type) as decorator for no-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        noValueParam
+    }
+
+    const x = new X()
+    x.noValueParam = function *(initial: number) {
+        this.a = initial // this runs in action
+        try {
+            this.a = yield delay(100, 5, true) // and this as well!
+            yield delay(100, 0)
+            this.a = 4
+        } catch (e) {
+            this.a = e
+        }
+        return this.a
+    }
+
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.noValueParam(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support canceling the effects when call canceler", t => {
+    mobx.useStrict(true)
+    const values: any[] = []
+    const x = mobx.observable({ a: 1 })
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    let canceler
+    const f = utils.asyncActionWithCancel(cancel => canceler = cancel)(function*(initial: number) {
+        x.a = initial // this runs in action
+        try {
+            x.a = yield delay(1000, 5, true) // and this as well!
+        } catch (e) {
+            x.a = e
+        }
+        return x.a
+    })
+
+    setTimeout(() => {
+        f(2)
+        setTimeout(() => {
+            canceler()
+            t.is(x.a, 2)
```

Comment:
When canceling would have failed, `x.a` would have been `5` here, correct?

Reply:
You are right!
As the cancel is just able to cancel the changes that have not happened. In other words, it's specially for **async** operation.

```
    asyncActionWithCancel(canceler => someJobCancel = canceler)(function someJob* () {
        var x = yield 2
    })
    someJobCancel()
    console.log(x) // 2
```

It's possible to revert the value by designing a special structure to make a time travel.(high complexity)